Design-Docs/standardizing-cluster-infrastructure.md

```diff
@@ -0,0 +1,277 @@
+
+There are various components, tools, and possible configurations that play a role in the pre-cluster state. This is a list that should serve as a reference:
+
+- Node problem detector (https://github.com/kubernetes/node-problem-detector): It depends on the OS because node errors do.
+- konnectivity-server ([https://kubernetes.io/docs/tasks/extend-kubernetes/setup-konnectivity](https://kubernetes.io/docs/tasks/extend-kubernetes/setup-konnectivity/)): kube-api-server proxy for traffic from control planes to cluster
+- Production-readiness: log-collector, log-rotator, pinning versions, DNS, NTP, ssh-hardening
+- Container Runtime (CRI-conform): cri-o, containerd
+- Low-Level Container Runtime (OCI-conform): crun, runc, runsc (gvisor), kata containers. With the help of runtime objects, this could be chosen directly in Kubernetes. Kata containers, for example, don’t work with virtualized systems without nested KVM support.
+- cgroups V2 vs V1, croupfs vs systemd
+- eBPF support: requires kernel >5.10
+- iptables requirements: depends on whether kube-proxy is used, which is not the case in some modern systems
+- SSH hardening
+- Firewall: local software firewall, externally or via cilium eBPF firewall
+
+##### Layer 2: Cluster state
+
+Cluster state consists of standardized Cluster API objects, such as `ClusterClass`, and other life cycle objects, e.g. `MachineDeployment` and `MachineHealthCheck`. Additionally, there are provider-specific objects that exist in every provider integration, like `ProviderCluster` and `ProviderMachineTemplate`, as well as other objects.
+
+SCS has to define the inputs for those objects in each Cluster Stack in such a way, that each provider can implement it.
+
+The most important standardization, but also the one that restricts the options to define clusters the most, is related to how the Kubernetes system is supposed to look like. Note that the Kubernetes system is heavily influenced by the components chosen for the pre-cluster state.
+
+The first decision that has to be taken is to select a bootstrap provider: Kubeadm or Talos. Talos would require a node image with Talos OS.
+
+A note here on the choice of bootstrap provider: the CAPI control-plane-provider and bootstrap-provider depend on this and have to be either from Kubeadm or Talos. This has to be supported by the infrastructure provider as well.
+
+A specific Kubernetes version has to be chosen as well, as other components depend on it.
+
+**What do we consider part of every cluster stack?**
+
+We recommend choosing these components for every stack to make sure that the cluster works well, is secure and that the maintainability of the cluster stacks is better. As all of the settings that we propose differ a lot between Talos and Kubeadm, we choose the latter one, as it is more common.
+
+In Kubeadm there are some settings that we would consider must-haves:
+
+- cloud-init
```

Comment:
What is cloud-init in the kubeadm context? I only know of [this cloudinit](https://cloudinit.readthedocs.io/), but that seems unrelated in this stage?

Reply:
We are using kubeadm without cloud-init since we miss it at this stage at gridscale.

I would be fine with missing something and therefore not be able to certify but this again looks like an implementation detail.

Reply:
@horazont the bootstrap provider kubeadm can generate output files for cloud-init and ignition. That's the relation between the two!